### DISCUSSION TITLE
Would like to run multiple API queries concurrently, but they run sequentially

### DISCUSSION THREAD

**Discussion by svet-b** [grafana/grafana-infinity-datasource#560] [Q&A]
I was surprised to find that if multiple queries are defined, they run sequentially - and am wondering if there's a way to make them run concurrently.

Namely, I have a situation that's similar to the pagination question [here](https://github.com/yesoreyeram/grafana-infinity-datasource/discussions/488#discussioncomment-4458422) - and I'm using the suggested workaround, where I define multiple API queries, one for each page of data. I observed that the Grafana backend waits for each of the API queries to complete before executing the next one - and returning the combined data once the last one completes. If each query takes ~15 seconds to complete, loading 4-5 pages of data takes over a minute - resulting in quite a poor user experience. Ideally the separate queries would run concurrently.

I'm not 100% sure whether the sequential execution is inherent to the Infinity plugin or some other part of the Grafana stack - but my tests so far indicate it's probably specific to Infinity. Any advice appreciated!

**Comment by yesoreyeram** [comment]
Yea.. that's unfortunately plugin as seen here https://github.com/yesoreyeram/grafana-infinity-datasource/blob/b15f378a932b18673bf556bdb209780d49a3ebdc/pkg/pluginhost/handler_querydata.go#L22 but i think it is easy to fix and improve in coming versions..

**Reply by svet-b** [reply]
Great, thanks for the pointer - that's super helpful! My Golang skills are rudimentary but I'll see if I can come up with a neat concurrent version of it.

**Reply by svet-b** [reply]
I mean...who needs coding skills when there's ChatGPT 😂
![Screenshot 2023-03-21 at 3 00 28 PM](https://user-images.githubusercontent.com/12831860/226630086-60118b9b-96a5-4e50-bd2b-443ae786da7e.png)
(kind of doubt this would work out-of-the-box, but let's see...)

**Reply by yesoreyeram** [reply]
Almost you are there..

**Reply by svet-b** [reply]
Hm...maybe close, but no banana it seems. I got something that should work in principle - and it does if the "worker function" is something trivial like `time.Sleep(2 * time.Second)` - in that case the datasource returns in two seconds regardless of the number of "queries" that are theoretically triggered in the loop. But as soon as I actually try to run `QueryData()` instead of a simple sleep, then each of the queries is run in series again. So something somewhere seems to be blocking, and I'm not sure what. Not sure if I'm missing something simple (or complicated), or if you have any pointers?

[other things I tried in an attempt to get parallel execution: instantiate a dedicated client for each query; discard the query result rather than file it into `response.Responses[q.RefID]`]

**Reply by svet-b** [reply]
An excerpt from ChatGPT's mega-generic (but helpful for me) analysis of the potential blockers:
1. The function `QueryData` could be using a shared resource in a way that causes contention. If `QueryData` is accessing a shared resource, such as a file or database, in a way that causes contention, then the goroutines may be forced to execute in series.
2. The function `QueryData` could be using a single-threaded library or API. If `QueryData` is using a library or API that is single-threaded, then it may not be possible to execute the function in parallel.

I suspect it's the former, but not sure what exactly (if not the client instance)...

**Reply by svet-b** [reply]
Last update: I went ahead and (to the best of my ability) created clones/deep copies of the `ctx`, `infinity.Client`, and `req` structures before passing them to the `QueryData()` function - in the hope that this would avoid any deadlocks due to shared resources, and would allow the separate requests to run concurrently. Unfortunately I'm still getting sequential execution.

**Reply by yesoreyeram** [reply]
great progress btw. currently I am busy with other priorities and not focusing on infinity for couple of weeks. Once back, I will take a look if this needs still attention.

**Reply by svet-b** [reply]
Thanks! I was also away from this for a bit but about to look into it again. If you have any tips they would be hugely appreciated!

**Reply by yesoreyeram** [reply]
Hi @svet-b - Quick update. I am still putting this on hold as my priorities changed and wanted to do this as part of paginated results support in v2.

**Reply by svet-b** [reply]
Hi @yesoreyeram thanks for the response. I'm just back form parental leave and checking this out. Congratulations on the recent v2 release, looks like a lot of great work went into that! Including pagination of course. Out of curiosity, should I expect paginated queries to now run in parallel? I haven't had the chance to play around with the updated plugin yet, but plan to in the next couple of weeks - so just wanted to know what to expect.

**Reply by yesoreyeram** [reply]
Hiya.. unfortunately parallel queries not a thing yet. I am yet to finish some different paginations mechanism which requires extraction of the token and passing to next request. Once I figure out clean way of doing that, I am thinking of making parallel requests. 

Also working on server side transformation which also relevant to order of requests and response.  

Making parallel queries is not a hard thing it seems but I would like to wait and proceed after the above improvements. For better scalability. 

Hope this makes sense

**Reply by yesoreyeram** [reply]
Thank you for the wishes and wishes to you for your fatherhood.

**Reply by svet-b** [reply]
Thanks for the clarification @yesoreyeram, makes sense. Will look out for further updates. Thank you also for the kind words.
